### PR TITLE
Remotecommand test flake cleanup

### LINKED ISSUE #121233: [Flaking Test] UT TestWebSocketClient for closed connection
### Which jobs are failing?

https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-unit

### Which tests are failing?

kubernetes/staging/src/k8s.io/client-go/tools/remotecommand TestWebSocketClient_TextMessageTypeError 

### Since when has it been failing?

not very sure as I was noticing this while working on #121213 

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit

### Reason for failure (if possible)

```
(base) ➜  remotecommand git:(e2e_3085_kep) go test -v -run TestWebSocketClient_TextMessageTypeError -count 20
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.547910   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.549109   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.549676   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.552227   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:51.553229   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:57:56.549378   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59680->127.0.0.1:59679: use of closed network connection
E1014 19:57:56.549381   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59682->127.0.0.1:59681: use of closed network connection
E1014 19:57:56.549811   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59684->127.0.0.1:59683: use of closed network connection
E1014 19:57:56.552763   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59686->127.0.0.1:59685: use of closed network connection
E1014 19:57:56.553394   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59688->127.0.0.1:59687: use of closed network connection
E1014 19:58:01.562592   14819 websocket.go:480] Websocket Ping failed: write tcp 127.0.0.1:59690->127.0.0.1:59689: write: broken pipe
    websocket_test.go:928: expect stream to be closed after connection is closed.
--- FAIL: TestWebSocketClient_TextMessageTypeError (30.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.560619   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.01s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.561513   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.562235   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.563078   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.563886   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.564797   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.565428   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.567187   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.568615   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.570444   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.01s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.581075   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.581634   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.582861   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
=== RUN   TestWebSocketClient_TextMessageTypeError
E1014 19:58:21.583384   14819 v2.go:150] unexpected message type: 1
--- PASS: TestWebSocketClient_TextMessageTypeError (0.00s)
FAIL
exit status 1
FAIL	k8s.io/client-go/tools/remotecommand	30.192s
```

### Anything else we need to know?

to reproduce:
cd kubernetes/staging/src/k8s.io/client-go/tools/remotecommand
go test -v -run TestWebSocketClient_TextMessageTypeError -count 50 ( or higher)

### Relevant SIG(s)

/sig api-machinery

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug
/kind cleanup
/kind flake

#### What this PR does / why we need it:

Fixes https://github.com/kubernetes/kubernetes/issues/121233

Tightens up race conditions around simultaneous stream creation and connection teardown

```release-note
NONE
```

/assign @seans3
/sig cli apimachinery

### CODE REVIEW

**Comment by seans3** [comment]
Fixes https://github.com/kubernetes/kubernetes/issues/121233

**Comment by liggitt** [comment]
realized these flakes are also impacting 1.29, opened cherry-pick in https://github.com/kubernetes/kubernetes/pull/124172

File: staging/src/k8s.io/client-go/tools/remotecommand/websocket.go [hunk 1]
```diff
@@ -187,6 +187,9 @@ type wsStreamCreator struct {
 	// map of stream id to stream; multiple streams read/write the connection
 	streams   map[byte]*stream
 	streamsMu sync.Mutex
+	// setStreamErr holds the error to return to anyone calling setStreams.
+	// this is populated in closeAllStreamReaders
+	setStreamErr error
```
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/remotecommand/websocket.go:192]: interesting and subtle bug

**Pull request merged**